azure-iot-device/tests/iothub/shared_client_tests.py

```diff
@@ -743,11 +891,323 @@ def test_sastoken_ttl(self, client_class, x509):
+        sastoken = st.NonRenewableSasToken(sas_token_string)
+        token_uri_pieces = sastoken.resource_uri.split("/")
+        device_id = token_uri_pieces[2]
+        return device_id
+
+    @pytest.fixture
+    def hostname(self, sas_token_string):
+        # NOTE: This is kind of unconventional, but this is the easiest way to extract the
+        # hostname from a sastoken string
+        sastoken = st.NonRenewableSasToken(sas_token_string)
+        token_uri_pieces = sastoken.resource_uri.split("/")
+        hostname = token_uri_pieces[0]
+        return hostname
+
+    @pytest.fixture
+    def sas_client(self, client_class, mqtt_pipeline, http_pipeline, sas_token_string):
+        """Client configured as if using user-provided, non-renewable SAS auth"""
+        sastoken = st.NonRenewableSasToken(sas_token_string)
+        token_uri_pieces = sastoken.resource_uri.split("/")
+        hostname = token_uri_pieces[0]
+        device_id = token_uri_pieces[2]
+        sas_config = IoTHubPipelineConfig(hostname=hostname, device_id=device_id, sastoken=sastoken)
+        mqtt_pipeline.pipeline_configuration = sas_config
+        http_pipeline.pipeline_configuration = sas_config
+        return client_class(mqtt_pipeline, http_pipeline)
+
+    @pytest.fixture
+    def new_sas_token_string(self, hostname, device_id):
+        # New SASToken String that matches old device id and hostname
+        uri = "{hostname}/devices/{device_id}".format(hostname=hostname, device_id=device_id)
+        signature = "AvCQCS7uVk8Lxau7rBs/jek4iwENIwLwpEV7NIJySc0="
+        new_token_string = "SharedAccessSignature sr={uri}&sig={signature}&se={expiry}".format(
+            uri=urllib.parse.quote(uri, safe=""),
+            signature=urllib.parse.quote(signature, safe=""),
+            expiry=int(time.time()) + 3600,
+        )
+        return new_token_string
+
+    @pytest.mark.it(
+        "Creates a new NonRenewableSasToken and sets it on the PipelineConfig, if the new SAS Token string matches the existing SAS Token's information"
+    )
+    def test_updates_token_if_match_vals(self, sas_client, new_sas_token_string):
+
+        old_sas_token_string = str(sas_client._mqtt_pipeline.pipeline_configuration.sastoken)
+
+        # Update to new token
+        sas_client.update_sastoken(new_sas_token_string)
+
+        # Sastoken was updated
+        assert (
+            str(sas_client._mqtt_pipeline.pipeline_configuration.sastoken) == new_sas_token_string
+        )
+        assert (
+            str(sas_client._mqtt_pipeline.pipeline_configuration.sastoken) != old_sas_token_string
+        )
+
+    @pytest.mark.it(
+        "Raises a ClientError if the client was created with an X509 certificate instead of SAS"
+    )
+    def test_created_with_x509(self, mocker, sas_client, new_sas_token_string):
+        # Modify client to seem as if created with X509
+        x509_client = sas_client
+        x509_client._mqtt_pipeline.pipeline_configuration.sastoken = None
```

Comment:
>x509_client._mqtt_pipeline.pipeline_configuration.sastoken = None [](start = 8, length = 65)

I don't see where pipeline_configuration.sastoken gets set to None in the product code.

Reply:
azure.iot.device.common.pipeline.BasePipelineConfig default arg in __init__

---
In reply to: [506704515](https://github.com/Azure/azure-iot-sdk-python/pull/692#discussion_r506704515) [](ancestors = 506704515)